src/Processors/QueryPlan/DistinctStep.cpp

```diff
@@ -140,15 +140,19 @@ void DistinctStep::transformPipeline(QueryPipelineBuilder & pipeline, const Buil
 
                 if (distinct_sort_desc.size() < columns.size())
                 {
-                    pipeline.addSimpleTransform(
-                        [&](const Block & header, QueryPipelineBuilder::StreamType stream_type) -> ProcessorPtr
-                        {
-                            if (stream_type != QueryPipelineBuilder::StreamType::Main)
-                                return nullptr;
-
-                            return std::make_shared<DistinctSortedTransform>(
-                                header, distinct_sort_desc, set_size_limits, limit_hint, columns);
-                        });
+                    if (DistinctSortedTransform::isApplicable(pipeline.getHeader(), distinct_sort_desc, columns))
```

Comment:
Add tests with EXPLAIN to check that `DistinctSorted` is not used in non-applicable cases